lib/src/strings.dart

```diff
@@ -94,38 +95,103 @@ String _tryYamlEncodeSingleQuoted(String string) {
+
+    if (trimmedSplit.length > 1) {
+      /// Try folding to match specification:
+      /// * https://yaml.org/spec/1.2.2/#65-line-folding
+      trimmedString =
+          trimmedSplit.foldIndexed(first, (index, previous, current) {
+        if (index == 0) return previous;
+
+        /// If initially empty, this line holds only `\n`. It's okay. Also, if
+        /// this line is just an empty space.
+        ///
+        /// See https://yaml.org/spec/1.2.2/#64-empty-lines
+        if (current.trim().isEmpty) return previous + lineEnding + current;
+
+        var updated = current;
+
+        /// For consecutive non-empty lines, we add a `\n` with a space.
+        ///
+        /// We need to ensure the ACTUAL previous string was non-empty
+        if (trimmedSplit[index - 1].trim().isNotEmpty) {
+          /// We remove indent and check if the string starts with a space.
+          ///
+          /// Apply `\n` for `foo\nbar` but not `foo\n bar`.
+          if (!current.replaceFirst(indent, '').startsWith(' ')) {
+            updated = lineEnding + updated;
+          }
+        }
 
-  if (removedPortion.contains('\n')) {
-    result = '>+\n${' ' * indentation}';
-  } else {
-    result = '>-\n${' ' * indentation}';
+        /// Apply `\n` by default if no operation occurred
+        return previous + lineEnding + updated;
+      });
+    } else {
+      trimmedString = first;
+    }
   }
 
-  /// Duplicating the newline for folded strings preserves it in YAML.
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result +
-      trimmedString.replaceAll('\n', lineEnding * 2 + ' ' * indentation) +
-      removedPortion;
+  return literalPrefix + trimmedString + stripped;
 }
 
 /// Generates a YAML-safe literal string.
 ///
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeLiteral(
-    String string, int indentation, String lineEnding) {
-  final result = '|-\n$string';
+String _tryYamlEncodeLiteral(String string, int indentSize, String lineEnding) {
+  final (literalPrefix, indent, trimmedString, stripped) = _prepareBlockString(
+    '|',
+    string,
+    indentSize,
+    lineEnding,
+  );
+
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return literalPrefix + indent + trimmedString + stripped;
+}
 
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
+/// Prepares a string to be encoded either via [_tryYamlEncodeLiteral] or
+/// [_tryYamlEncodeFolded] when in [CollectionStyle.BLOCK].
+(
+  String literalPrefix,
+  String indent,
+  String trimmedString,
+  String strippedString,
+) _prepareBlockString(
+    String blockIndicator, String string, int indentSize, String lineEnding) {
```

Comment:
btw, feel free to insert assertions, they are practically free when a program is compiled in release mode. And a few sanity checks like this can only help catch bugs :D


```suggestion
    String blockIndicator, String string, int indentSize, String lineEnding) {
    assert(indentSize >= 0);
```

Reply:
I realised there is no need to `assert`. We can pass in the `indent` and just apply to all strings that appear after `\n`